src/layers/tile/util.rs

```diff
@@ -11,47 +11,28 @@ pub(crate) fn parse_data_line(
     tilesets: &[MapTilesetGid],
 ) -> Result<Vec<Option<LayerTileData>>, TiledError> {
     match (encoding.as_deref(), compression.as_deref()) {
-        (Some("base64"), None) => {
-            return parse_base64(parser).map(|v| convert_to_tiles(&v, tilesets))
-        }
-        (Some("csv"), None) => return decode_csv(parser, tilesets),
-        (Some(_), None) => {
-            return Err(TiledError::InvalidEncodingFormat {
-                encoding,
-                compression,
-            })
-        }
-        (Some(e), Some(c)) => match (e, c) {
-            ("base64", "zlib") => {
-                return parse_base64(parser)
-                    .and_then(decode_zlib)
-                    .map(|v| convert_to_tiles(&v, tilesets))
-            }
-            ("base64", "gzip") => {
-                return parse_base64(parser)
-                    .and_then(decode_gzip)
-                    .map(|v| convert_to_tiles(&v, tilesets))
-            }
-            #[cfg(feature = "zstd")]
-            ("base64", "zstd") => {
-                return parse_base64(parser)
-                    .and_then(decode_zstd)
-                    .map(|v| convert_to_tiles(&v, tilesets))
-            }
-            _ => {
-                return Err(TiledError::InvalidEncodingFormat {
-                    encoding,
-                    compression,
-                })
-            }
-        },
-        _ => {
-            return Err(TiledError::InvalidEncodingFormat {
-                encoding,
-                compression,
-            })
-        }
-    };
+        (Some("csv"), None) => decode_csv(parser, tilesets),
+
+        (Some("base64"), None) => parse_base64(parser).map(|v| convert_to_tiles(&v, tilesets)),
+        (Some("base64"), Some("zlib")) => parse_base64(parser)
+            .map(|data| std::io::Cursor::new(data))
```

Comment:
I still don't really like this duplication of `parse_base64` and `convert_to_tiles` for each base64-encoded data, but in trying to make it shorter we of course get duplicated error handling and I had to make a clone of the encoding and compression for the error:
```rust
    match (encoding.as_deref(), compression.as_deref()) {
        (Some("csv"), None) => decode_csv(parser, tilesets),
        (Some("base64"), c) => {
            parse_base64(parser).and_then(|bytes| {
                match c {
                    None => Ok(bytes),
                    Some("zlib") => process_decoder(libflate::zlib::Decoder::new(&bytes[..])),
                    Some("gzip") => process_decoder(libflate::gzip::Decoder::new(&bytes[..])),
                    #[cfg(feature = "zstd")]
                    Some("zstd") => process_decoder(zstd::stream::read::Decoder::with_buffer(&bytes[..])),
                    _ => Err(TiledError::InvalidEncodingFormat {
                        encoding: encoding.clone(),
                        compression: compression.clone(),
                    }),
                }
            }).and_then(|v| Ok(convert_to_tiles(&v, tilesets)))
        },
        _ => Err(TiledError::InvalidEncodingFormat {
            encoding,
            compression,
        }),
    }
```

So, I wouldn't suggest to go with the above.

However, it seems we can at least get rid of the `std::io::Cursor`, since we can just do:
```rust
|data| process_decoder(libflate::zlib::Decoder::new(&data[..]))
```
Or is there some disadvantage to doing that?

Reply:
Didn't know `Read` was implemented for `&[u8]`, will change later

And yes, that base64 duplication is annoying but as far as i know there isn't any clean way of changing it